Update section 'Batch Run' of introductory tutorial

jonahaag opened the pull request:
As discussed in #1109, I updated the introductory tutorial to reflect the new `batch_run` function. 
For this I edited the corresponding section in the Juypter Notebook and generated the reStructuredText file from that. This also automatically generated the images in the `intro_tutorial_files` folder based on the results that I received when testing.

Note: I have not yet run `make html` to build the docs as I wasn't too sure about the exact procedure and wanted to discuss the changes first.

TODO:

- [x] State the names of the arguments of `batch_run` when talking about them. -> Should be solved by https://github.com/projectmesa/mesa/pull/1119/commits/04b01c6bc3097c8ec6ad740d33f21875b0f1883f
- [x] Elaborate on fixed and variable parameters -> Should be solved by https://github.com/projectmesa/mesa/pull/1119/commits/04b01c6bc3097c8ec6ad740d33f21875b0f1883f
- [x] State that processors are used to run sweep in parallel -> Should be solved by https://github.com/projectmesa/mesa/pull/1119/commits/04b01c6bc3097c8ec6ad740d33f21875b0f1883f
- [ ] Rename `nr_processes` -> different PR
- [ ] Build the docs

rht commented:
I just realized that the `i_steps` param in `batch_run` is not yet documented in the docstring, but is documented in the helper internal functions. Also, I wonder if it should be called `data_collection_period`, because `i_steps` is too cryptic.

jonahaag commented:
Btw, setting `max_steps=100` results in the episodes to be of length 101 since there [is a `<=` in the while loop condition](https://github.com/projectmesa/mesa/blob/0ae8410098934e7cf363784fae74f12183003202/mesa/batchrunner.py#L163) and episodes start with `step=0`.

tpike3 commented on docs/tutorials/intro_tutorial.rst:
```diff
@@ -879,273 +895,204 @@ True indefinitely.
-dictionaries of model- and agent-level reporters to collect. Unlike the
-DataCollector, it won't collect the data every step of the model, but
-only at the end of each run.
+We call a batch run with a model class to run, and a dictionary with the
```
I would turn batch run into the function so 'We call ```batch_run```'

tpike3 commented on docs/tutorials/intro_tutorial.rst:
```diff
@@ -879,273 +895,204 @@ True indefinitely.
+iterable. Further, we can specify how many model instantiations to
+create and run at each combination of parameter values, and how many
+steps to run each instantiation for. We can also set the number of
+processes to be run at the same time (default ``None`` uses all
```
I would change 'processes' to 'processors'

tpike3 commented on docs/tutorials/intro_tutorial.rst:
```diff
@@ -879,273 +895,204 @@ True indefinitely.
 
-Now, we can set up and run the BatchRunner:
+Since for the latter changes at each time step might be interesting, we
+set ``i_steps=1``.
```
Seeing this in the tutorial I concur with @rht ```i_steps``` is cryptic if this can be changed to ```data_collection_period``` that would be great

jonahaag replied:
Should I open a new PR for this or just add these changes here?

rht replied:
I think the `i_steps` -> `data_collection_period` rename should be done in a separate PR. 2 smaller PRs are easier to review, and faster to be merged than 1 big PR altogether.

jonahaag replied:
Just opened a new PR, will rebase and update this one once the new one is merged.

tpike3 commented on docs/tutorials/intro_tutorial.rst:
```diff
@@ -879,273 +895,204 @@ True indefinitely.
+time) at the 100th step of each episode and then scatter-plot the values
+for the Gini coefficient over the the number of agents. Notice there are
+five values for each population size since we set ``iterations=5`` when
+calling the batch run.
```
Excellent! Well done!

jonahaag replied:
Thanks for the nice feedback!

tpike3 reviewed (changes requested):
@jonahaag  Really nicely done with the Pandas and the data, very nice. I have a couple nitpicky requests for one parargraph but other than that it looks great! Thanks!

jonahaag commented:
I rebased, replaced `i_steps` with `data_collection_period`, tested if everything still works as expected, generated the rst file again, and finally made some small manual changes to the latter for proper code highlighting (see the [GitHub preview](https://github.com/jonahaag/mesa/blob/update-intro-tutorial/docs/tutorials/intro_tutorial.rst)).

rht commented:
> The batchrunner also requires an additional variable self.running for the MoneyModel class. 

Should be 'batch runner'.

rht commented on docs/tutorials/intro_tutorial.rst:
```diff
@@ -58,21 +58,21 @@ To install Mesa, simply:
 .. code:: bash
 
-       $ pip install -r https://raw.githubusercontent.com/projectmesa/mesa/main/examples/boltzmann_wealth_model/requirements.txt
+       $ pip install -r https://raw.githubusercontent.com/projectmesa/mesa/main/examples/boltzmann_wealth_model/requirements.txt  
```
Nit: is there any rendered change s a result of the 2 extra whitespaces?

rht commented on docs/tutorials/intro_tutorial.rst:
```diff
@@ -151,78 +155,82 @@ model uses, and see whether it changes the model behavior. This may not
             super().__init__(unique_id, model)
             self.wealth = 1
-
+    
```
Unnecessary extra whitespaces.

rht commented on docs/tutorials/intro_tutorial.rst:
```diff
@@ -151,78 +155,82 @@ model uses, and see whether it changes the model behavior. This may not
                 a = MoneyAgent(i, self)
                 self.schedule.add(a)
-
+    
```
Same here.

rht commented on docs/tutorials/intro_tutorial.rst:
```diff
@@ -239,25 +247,26 @@ Now we just need to have the agents do what we intend for them to do:
             super().__init__(unique_id, model)
             self.wealth = 1
-
+    
```
Same here.

rht commented on docs/tutorials/intro_tutorial.rst:
```diff
@@ -268,109 +277,110 @@ With that in mind, we rewrite the agent ``step`` method, like this:
     # For a jupyter notebook add the following line:
     %matplotlib inline
-
+    
```
Here.

rht commented on docs/tutorials/intro_tutorial.rst:
```diff
@@ -268,109 +277,110 @@ With that in mind, we rewrite the agent ``step`` method, like this:
     # The below is needed for both notebooks and scripts
     import matplotlib.pyplot as plt
-
+    
```
Here.

rht commented on docs/tutorials/intro_tutorial.rst:
```diff
@@ -879,273 +895,208 @@ True indefinitely.
-DataCollector, it won't collect the data every step of the model, but
-only at the end of each run.
+We call ``batch_run`` with a model class to run, and a dictionary with
+the model parameters where each of the values can either be fixed or an
```
I think we need to elaborate on how to specify the parameters in the case of fixed vs iterable.

rht replied:
The older documentation already explains it succinctly, because of the clear separation between fixed and varying parameters.

tpike3 replied:
I agree as this often becomes a sticky point for students... just a thought on how to succinctly convey and I by no means tied to this. 

_We call ``batch_run`` with a model class to run, and a dictionary with the model parameters where each of the values can either be fixed ( e.g. {"height": 10, "width": 10}) or an iterable  (e.g. "num_agents": range(10, 500, 10) )_

rht commented on docs/tutorials/intro_tutorial.rst:
```diff
@@ -879,273 +895,208 @@ True indefinitely.
+iterable. Further, we can specify how many model instantiations to
+create and run at each combination of parameter values, and how many
+steps to run each instantiation for. We can also set the number of
+processors to be run at the same time (default ``None`` uses all
```
You haven't said that those processors are used to run the sweep in parallel.

tpike3 replied:
Yes, let's highlight the multiprocessing point as multiprocessing can often get temperamental depending on the OS. 

Again a thought, but feel free to not use or make better:

_Further, we can specify how many model instantiations to create and run at each combination of parameter values, and how many steps to run in each instantiation. By default, ```batch_run``` will use every processor so the different model variations are run in parallel. Multiprocessing does make debugging challenging. If your parameter sweeps are resulting in unexpected errors set ```nr_processes = 1```.  ```batch_run``` can also set the number of steps that data is collected with with keyword argument  ```data_collection_period```._

rht commented on docs/tutorials/intro_tutorial.rst:
```diff
@@ -879,273 +895,208 @@ True indefinitely.
-    run_data.head()
-    plt.scatter(run_data.N, run_data.Gini)
+To further analyze the return of the ``batch_run`` function, we convert
+the list of dictionaries to a pandas data frame and print its keys.
```
*Pandas DataFrame.

jonahaag replied:
> To do this, we again filter our large data frame, this time with a fixed number of agents and only for a specific iteration of that population. To print the results, we convert the filtered data frame to a string specifying the desired columns to print.

For this part there is no need to rename, right?

rht replied:
Hmm yeah, I think there is no need to rename the rest of the usage. introducing DataFrame with its proper name once, is sufficient.

But I do need to say that, Pandas doc always refers their object as DataFrame (see https://pandas.pydata.org/docs/search.html?q=data+frame), even though the concept "data frame" exists in R, etc.

jonahaag replied:
I'll leave it as is for the moment

rht commented:
- Sorry for the nit, but there are several extraneous whitespaces (I found them in the GitHub diff UI). I thought there are few of them and hence I commented on each of them, but there are more.
- I think the section in the tutorial that describes the parameters in `batch_run` can also say the parameter names, e.g. `nr_processes` while describing about `nr_processes`.
- This is yet another variable renaming: I think `nr_processes` can either be expanded to `number_processes` or `number_of_processes` (I don't see the point of shortening 'number' to 'nr'), or use Python `multiprocessing`'s naming, https://docs.python.org/3/library/multiprocessing.html#multiprocessing.pool.ThreadPool, which is just `processes`. This can be fixed after this PR is merged

jonahaag commented:
Thanks for the review! 

I guess the whitespaces were a result of me using the `jupyter nbconvert --to rST *.ipynb` command to autogenerate the rst-file as suggested in [docs/README.md](https://github.com/projectmesa/mesa/tree/main/docs#readme). Same goes for all the `'` being replaced by `´` and some new blank lines, the latter being part of the code cells of the notebook already but not of the rst-file.

TODO:

- [ ] State the names of the arguments of `batch_run` when talking about them. 
- [ ] Elaborate on fixed and variable parameters
- [ ] State that processors are used to run sweep in parallel
- [ ] Rename `nr_processes` -> different PR

Maybe it makes sense to use a list describing each argument instead of plain text since there are quite a few here (similar to the docstrings but more elaborate). Something like: 
> `batch_run` has the following arguments:
> - `model_cls`: The model class that is used for the batch run. 
> - `parameters`: A dictionary containing all the parameters of the model class and desired values to use for the batch run as key-value pairs. Each value can either be fixed or an iterable, e.g. a list. `batch_run` will then generate all possible parameter combinations based on this dictionary and run the model `iterations` times for each combination. 
> - `nr_processes`: Number of processors used to run the sweep in parallel. Optional int, defaults to `None` to use all available processors.
> - `iterations`: The number of iterations to run each parameter combination for. Optional int, defaults to 1. 
> - `data_collection_period`: The length of the period / the number of steps after which the model and agent reporters collect data. Optional, defaults to -1, i.e. only collect data at the end of each episode.
> - `max_steps`: The maximum number of time steps after which the model halts. An episode does either end when `self.running` of the model class is set to `False` or when `model.schedule.steps == max_steps` is reached. Optional int, defaults to 1000. 
> - `display_progress`: Display the batch run progress. Optional bool, defaults to `True`.

What do you think about this?

rht commented:
Yeah, that list with "`arg_name`: description" is more concise an informative to me.

tpike3 commented:
I concur the list with the arguments is important. I honestly was about to argue against based on information overload for an intro tutorial but....... each of those parameters is pretty important.

jonahaag commented:
Any suggestions regarding the content of that list? This should address all of the three TODOs for this PR.

rht commented:
@tpike3  I see your point with the concern of information overload. But the original form with the series of sentences is more overwhelming to me as a reader.

>  Optional int, defaults to None to use all available processors.

This is too cryptic for an introduction? I'd say instead "Optional. If not specified, defaults to use all the available processors".

jonahaag commented:
I agree on the table being less overwhelming. I could also leave out some of the parameters and instead link to [batchrunner.py](https://github.com/projectmesa/mesa/blob/68c48e6d1de2a3b7d102a3530920b2a4f5cc08e7/mesa/batchrunner.py#L45) for the remaining ones, although, as @tpike3 said, all of them are pretty important (at least apart from `display_progress`).

I also prefer the suggested description for `nr_processes`. Probably makes sense to omit the data types for all of them, I think those are pretty obvious given the default values and descriptions.

Would result in:
> We call `batch_run` with the following arguments:
> - `model_cls`: The model class that is used for the batch run. 
> - `parameters`: A dictionary containing all the parameters of the model class and desired values to use for the batch run as key-value pairs. Each value can either be fixed or an iterable, e.g. a list. `batch_run` will then generate all possible parameter combinations based on this dictionary and run the model `iterations` times for each combination. 
> - `nr_processes`: Number of processors used to run the sweep in parallel. Optional. If not specified, defaults to use all the available processors.
> - `iterations`: The number of iterations to run each parameter combination for. Optional. If not specified, defaults to 1. 
> - `data_collection_period`: The length of the period (number of steps) after which the model and agent reporters collect data. Optional. If not specified, defaults to -1, i.e. only at the end of each episode.
> - `max_steps`: The maximum number of time steps after which the model halts. An episode does either end when `self.running` of the model class is set to `False` or when `model.schedule.steps == max_steps` is reached. Optional. If not specified, defaults to 1000. 
> - `display_progress`: Display the batch run progress. Optional. If not specified, defaults to `True`.

tpike3 reviewed (changes requested):
@jonahaag Thanks again! This is great improvement and the nitpicky part is always the final step, but pays off as it dramatically reduces user issues.

jonahaag commented:
Ok, so would you stick with plain text or include the suggestions you made in the list that we discussed? I haven't committed a version with the list yet, since I wasn't sure whether we're going to opt for that.

rht commented:
@tpike3 did you change your mind about the list format, and why? The boundary between the parameters description are much clearer with the list format, and it is easier for the eye to look at if someone wants to quickly see again the description of a param.

tpike3 commented:
I did not. I think we should have the list, @jonahaag please include the list and please include the more verbose description as noted in the review. 

Make sense?

jonahaag commented:
Last 3 commits contain the following changes:

- Fix the citation reference of the two papers in intro_tutorial.rst
- Change to the list to describe arguments of `batch_run`
- Add some blank lines, indentation etc. to improve visual appearance of the list. Looked very packed the way it was initially.
- Include examples for fixed and variable parameters as suggested
- Include note for potential debugging issues using multiprocessing
- Remove sentence about the default value of `data_collection_period` from the subsequent paragraph, now explained in the list

As discussed, I would address the rename of `nr_processes` in a separate PR after this is merged. I think it's best to update the tutorial accordingly once we chose the name (so as part of the "rename PR").

rht commented:
For quick reference for those who want to see the rendered .rst: https://github.com/jonahaag/mesa/blob/update-intro-tutorial/docs/tutorials/intro_tutorial.rst

This PR LGTM.

tpike3 commented:
@jonahaag Awesome, I am ready to rebase and merge but could you do one admin thing for me and [squash the commits](https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/managing-commits/squashing-commits) into one commit. Thanks.

jonahaag commented:
Great, thanks again @rht and @tpike3! Is there anything for me do in terms of building the docs or is somebody going to take care of that?
